### ISSUE TITLE
Allow type inference for class members with autoinitializers and methods (use "var"/"auto" keywords)

### ISSUE THREAD

**Issue by Makeloft** [dotnet/csharplang#1181]
Allow type inference for class members with autoinitializers and methods (use "var"/"auto" keywords)

```
class SomeClass
{
    public static readonly SomeClass Instance = new SomeClass();
    public const SomeClass ConstInstance = new SomeClass();
    public List<string> StringsField = new List<strings>();
    public List<string> StringsProperty { get; set; } = new List<strings>();

    public string Cover(string str) => "[" + str + "]";
}

class SomeClass
{
    public static readonly var Instance = new SomeClass();
    public const var ConstInstance = new SomeClass();
    public var StringsField = new List<strings>();
    public var StringsProperty { get; set; } = new List<strings>();

    public var Cover(string str) => "[" + str + "]";
}

class SomeClass
{
    public static readonly var Instance = new SomeClass();
    public const var ConstInstance = new SomeClass();
    public var StringsField = new List<strings>();
    public auto StringsProperty { get; set; } = new List<strings>();

    public auto Cover(string str) => "[" + str + "]";
}
```

```
        public static IEnumerable<T> ToEnumerable<T>(this T singleItem)
        {
            yield return singleItem;
        }

        public static auto ToEnumerable<T>(this T singleItem) => yield return singleItem;
```

```
        public static IDictionary<TKey, TValue> AppendRange<TKey, TValue>(
            this IDictionary<TKey, TValue> target,
            IEnumerable<KeyValuePair<TKey, TValue>> source)
        {
            foreach (var item in source) target.Add(item);
            return target;
        }

        public static auto AppendRange<TKey, TValue>(
            this IDictionary<TKey, TValue> target,
            IEnumerable<KeyValuePair<TKey, TValue>> source)
        {
            foreach (var item in source) target.Add(item);
            return target;
        }
```

**Comment by Joe4evr** [comment]
This has been proposed many times before.
See also Target-typed `new`: #100 (which is the inverse of this). `var` fields are too hard to do [because there could be cycles and the compiler can't detect that](https://github.com/dotnet/csharplang/issues/100#issuecomment-303187152).

And please, in the future write more about what you're proposing instead of just putting in some code and calling it a day.

**Comment by Makeloft** [comment]
What about methods and properties? Is there hard to implement too?

**Comment by Makeloft** [comment]
Also Target-typed "new" is not usable for instances with parametric constructors.

`    public var StringsField = new List<strings>(capacity: 128) { "abc" };`

**Comment by huoyaoyuan** [comment]
As I know, the C# team has a flavor that "all public api should be write explicitly". Type of fields is part of it. Local variables are not public, so they introduced `var`.

**Comment by Makeloft** [comment]
But all of this members may have 'private' modifier too. And there is not public api in this case. :)

**Comment by yaakov-h** [comment]
You can’t infer for your `yield` example because yield-returning methods can either return an `IEnumerable<T>` or an `IEnumerator<T>`

**Comment by Makeloft** [comment]
Nice remark! Seems that suggested type inference is not suitable for yield-returning methods.

**Comment by jnm2** [comment]
Devil's advocate: While I think type inference would be a horrible mistake for any member visible outside the assembly, yield-returning methods could default to `IEnumerable<>`.

**Comment by jnm2** [comment]
It would be at least as bad as dynamically marking methods as pure based on implementation, if not worse. Spooky action at a distance and all that. You should be forced to make a conscious decision to change the public contract to match the change in implementation.

**Comment by Joe4evr** [comment]
> Also Target-typed "new" is not usable for instances with parametric constructors.

Says who?
```cs
public List<string> Strings { get; } = new(capacity: 128);
```

**Comment by CyrusNajmabadi** [comment]
> As I know, the C# team has a flavor that "all public api should be write explicitly".

This is not true.  The primary issue is one around implementation difficulty.  It turns out to be *really* hard with the existing compiler implementation to support this sort of thing.  The compiler has always been able to assume, and take advantage of, the types of members being known before any "code" (as opposed to signatures) have to be bound.  

Also, many performance optimizations are done based on this constraint of the language.  For example, in intellisense, we know that edits inside methods can not have any impact on type information in your system.  However, if methods can now have 'var' return types, then your edits can have widespread impact and such optimizations would have to be removed.  

I think many people on the LDM are very interesting/open to the idea of being able to use 'var' judiciously in more places to clean the langauge up and avoid all the type-redundancy you can see.  However, investigations into this have shown it is non-trivial and comes with very significant and serious engineering concerns.

**Comment by CyrusNajmabadi** [comment]
IMO, from an engineering perspective, 'target-typed new' is the far safer and simpler approach vs 'allow var in more cases'.  It accomplishes the goal of DontRepeatYourself, but keeps the compiler implementation much saner, where inference stays only in "code" and doesn't get elevated to "signatures" which would vastly complicate our architecture.

**Comment by Makeloft** [comment]
@Joe4evr thanks! But the main reason why I want to have left type inference for fields or properties related with workflow.

It will be very nice to have opportunity move local method variable to class level without serious string modifications, just Cut&Paste or Drag&Drop. Of course in both directions.

```
class AnyClass
{
    public void AnyMethod()
    {
        var anyVariable = new AnyType();
        ....
    }
}
```
then Cut&Paste
```
class AnyClass
{
        var anyVariable = new AnyType();

    public void AnyMethod()
    {
        ....
    }
}
```
compiled successfully without serious modifications.

**Comment by Makeloft** [comment]
@CyrusNajmabadi thanks for opinion!

I believe that good design of language should be maximally independent from compiler implementation.

From external point of view all this 'var'-features looks like syntax sugar. If we try to compile method with mismatching return value or other members with invalid initializers compiler shows the errors. What problems to replace token 'var' or 'auto' to inference return type tokens at the syntax tree?

I don't know details of compiler, but JetBrains ReSharper shows mismatched types at real time. Looks that it is possible to implement left side inference for different class members.

Sorry for my bold sayings!  :)

**Comment by CyrusNajmabadi** [comment]
> I believe that good design of language should be maximally independent from compiler implementation.

Ok :)
Language design team: we want feature x.
Compiler team: Ok, that will take us 20 person years to make the changes to make that possible.

Now what?

**Comment by CyrusNajmabadi** [comment]
>  If we try to compile method with mismatching return value or other members with invalid initializers compiler shows the errors. What problems to replace token 'var' or 'auto' to inference return type tokens at the syntax tree?

I pointed out the problem :)  It great increases compiler complexity.  *Today* for example, things like intellisense do not have to cause the world to get reanalyzed when a method body is edited.  This is a *very good thing*.  It means you can type lots of stuff in a method and we can keep intelllisense nice and fast and correct because it can cache and reuse the majority of the data that was computed when you made your original edit.  With 'var' returning methods that may no longer be the case.

Similarly, *today* the compiler can analyze easy method independently.  That's also a good thing (for example, for parallelism and perf), as well as for clean compiler design.  With 'var' returning methods that's no longer the case.  Now, when the compiler is analyzing one method, it may need to go analyze *another* method to figure out what it means.  The compiler currently doesn't have any facility for this, and it would introduce a data dependence between members.  This is something that can be done, but is non trivial.

Again, no one said it isn't possible.  The question is: how costly is this.  And smart language design means not going with features that have humongous costs compared to the end value they end up providing the user.

**Comment by CyrusNajmabadi** [comment]
Here's another example: today the compiler can do inheritance checking and validation (i.e. signature validation) *prior* to do any method body analysis.  This part of the compiler woudl now no longer work.  This introduces a circularity in teh compiler where previously there were nicely demarcated data flow paths.  i.e. in order to do inheritance, the compiler needs to compile method bodies.  But method bodies need to understand the inheritance of the type in order to properly do name lookup.

Again, this is solvable, but *very* complex.  It is made *more* complex by virtue of the fact that roslyn is a *highly concurrent* compiler.  i.e. the *expected* use mode for the compiler is that many threads are hitting it and making requests of it (for example, colorization may be trying to bind an identifier in a method to colorize it, while error squiggles may be causing inheritance to be bound in order to show error messages).  These requests happen simultaneously, and there's a *ton* of complexity in the compiler to make sure these operations can happen in parallel and always produce correct and consistent results.  Data dependencies are part of this and we make heavy use of how members and inheritance work today.

this is solvable, but difficult.  So the question is: does the feature provide enogh value given it's high cost.

**Comment by CyrusNajmabadi** [comment]
Finally, i would note again:

1. 'var for members': Allows you do not repeat types.  Very high cost.
2. 'new + inferred types': Allows you to not repeat types.  Lower cost.

So the question to me is: why would we do '1' when '2' accomplishes the same goal, at vastly lower cost?

**Comment by Makeloft** [comment]
@CyrusNajmabadi seems that in this situation compiler has fundamental limitations, possible, it may sense redesign compiler at all.  :)

Kidding, of course, but with part of true.

Looks in this cases that compiler and analizer transforming into legacy software with serious promblems and there is required step to create new generation of this tools.

**Comment by CyrusNajmabadi** [comment]
> seems that in this situation compiler has fundamental limitations

Yes.  All software does.  

> it may sense redesign compiler at all. :)

Redesigning the entire compiler to get this feature would be a bad usage of resources.  Especially since we can accomplish the *actual end goal* with a different language feature.

Seriously, i don't even understand how this would make sense:  Spend person decades to make it possible to type less types.  Or spend a few weeks/months to accomplish the same task by coming at it from a different angle.  Picking the former would be *bad software engineering*.  Worse yet, it would be punishing our users by forcing them to wait years and years to get this language feature (and others).

> Looks in this cases that compiler and analizer are transforming into legacy software with serious promblems 

No.  We've simply designed things in certain ways that have pros and cons, just like any piece of software.  These pros and cons are well understood and were made knowing that they would deliver significant advantages while also *not* really limiting us from being able to solve problems.  To repeat, we're not limited from solving the problem of "not repeating types".  We may be limited in terms of specific solutions.  

What i'm trying to convey to you is to get out of the mindset of expecting a *specific solution* to a problem.  Language design is not about specific solutions.  It's about asking "what problem are we trying to solve?  what resources would it take to solve it?" and coming up with a reasonable solution that fits within your constraints.  

For this problem, and our current compiler+ide architecture, 'var everywhere' would be quite costly.  So instead of going after that, we can offer alternative approaches that address the actual problem at far less cost.

**Comment by Makeloft** [comment]
@CyrusNajmabadi big thank for detailed answer! Of course, I understand your arguments and agree with them into business context.

But... For me developing is like art. And actually C# is one of best designed multiparadigmatic languages. I very like mathematical beauty of them. Sure that there is very important to save and improve this beauty for us and next generations of developers, if possible. )

**Comment by CyrusNajmabadi** [comment]
> And actually C# is one of best designed multiparadigmatic languages. 

I would also argue that hte Roslyn compiler is one of the best designed compilers out there.  Designed fomr the ground up to be flexible, highly concurrent, deeply immutable, extremely suitable for IDE/tooling uses, extremely fast,   etc. etc. etc.

But *fundamentally* with all engineering you have to make decisions and tradeoffs.  These design decisions have impact on what you can do and affect relative consts of things.  Just as it would be expensive to tell a javascript compiler "you now need to support compiling C++ as well".  Compilers will always make decisions based on what the language is, and where they think it will go.  If the language actually goes in an entirely different direction, then that can always affect a compiler.

> But... For me developing is like art. 

Great :)    Find beauty where you can.  Just don't expect that an actual product with millions of customers will bend to that desire instead of making deeply practical and pragmatic choices :)

**Comment by MgSam** [comment]
With all that being said, R# is able to infer return types for methods and members (and usually remain performant in doing so). I'd be interested to hear what they're doing differently.

**Comment by CyrusNajmabadi** [comment]
One of the most common things we hear (and see through telemetry) is resharper using tons of resources :)   I know from personal experience that it can add 2+ GB of memory when i'm using it for only medium sized projects.  

They've optimized in other directions.  It makes some things probably more possible for them, while potentially making other things more difficult.

**Comment by CyrusNajmabadi** [comment]
I use Rider :)   It isn't magically free from perf issues  :)

**Comment by MgSam** [comment]
@CyrusNajmabadi 🤷‍♂️ 

Memory is cheap. Developer time is many orders of magnitude more expensive. I think they've made the right optimizations. 😄

**Comment by jnm2** [comment]
> Developer time is many orders of magnitude more expensive.

Then why is ReSharper's responsiveness usually somewhere between mildly irritating and outright infeasible for work? This is on computers with beefy CPUs and gobs of RAM.

**Comment by MgSam** [comment]
@jnm2 This is definitely the wrong platform for this discussion, but 99% of the time I've never had a problem. And for something a bit less anecdotal- if it didn't work really well, it wouldn't be such a wildly successful plugin.

**Comment by jnm2** [comment]
My point is that they've obviously risked burning developer time, and actually done it in my case and with multiple other people I've talked to. Roslyn so far has not. So I find that the developer time argument cuts both ways.

**Comment by jnm2** [comment]
Someone else I know just started complaining about big ReSharper performance problems. Funny timing. :D

**Comment by CyrusNajmabadi** [comment]
>  I think they've made the right optimizations. 😄

And i think we've done the same :)

**Comment by Makeloft** [comment]
@CyrusNajmabadi 
Ha-ha! :)
Remember that hours of work compiler team are extra large optimization and million hours of developer work in the world in the feature times.

**Comment by CyrusNajmabadi** [comment]
Agreed.  Hence why it's better to work on bigger bang/buck features (like "target typed new") which give you the same sort of result, vs less advantageous features :)

**Comment by Makeloft** [comment]
@CyrusNajmabadi 
But current feature suggest not only fields and properties optimization. It is suggest common approach for all members (methods too) and this give greater advantages for developers at all.

I can describe why I don't like "target typed new". Main minus for me that when I move declarations from function to class level  I need do something else...

When there is not var I can just Cut&Paste and all works good
```
class AnyClass
{
    AnyType anyType = new AnyType();

    void Method()
    {
        AnyType anyType = new AnyType();
        ....
    }
}
```
With var it is annoying me now sometimes
```
class AnyClass
{
    AnyType anyType = new AnyType();

    void Method()
    {
        var anyType = new AnyType();
        ....
    }
}
```
But with using of "target typed new" it worth for me X2 times! And, of course, analyzers will suggest to use new syntax by default...
```
class AnyClass
{
    AnyType anyType = new();

    void Method()
    {
        var anyType = new AnyType();
        ....
    }
}
```
Really as developer I expect
```
class AnyClass
{
    var anyType = new AnyType();

    void Method()
    {
        var anyType = new AnyType();
        ....
    }
}
```
Sorry, and will feel better without "target typed new".
It is not bring real advantage for me because intellisence after "new" complete "AnyType".

Only if there will be done both directions type inference I will be glad to see them too.
class AnyClass
{
    AnyType anyType = new();
    var anyType = new AnyType();

    void Method()
    {
        var anyType = new AnyType();
        ....
    }
}
```

**Comment by CyrusNajmabadi** [comment]
> It is suggest common approach for all members (methods too)

But it really doesn't.  Methods would still need to use types for their parameters.  So you are going to have to spell our your types commonly, no matter what.  

> With var it is annoying me now sometimes in function to class direction

Sure that's def an inconsistency.

--

I'm not sure where to go from here.  I've explained that no one is opposed to the concept of this feature.  As i've said repeatedly:

> This is not true. The primary issue is one around implementation difficulty. It turns out to be really hard with the existing compiler implementation to support this sort of thing. The compiler has always been able to assume, and take advantage of, the types of members being known before any "code" (as opposed to signatures) have to be bound.
>
> Also, many performance optimizations are done based on this constraint of the language. For example, in intellisense, we know that edits inside methods can not have any impact on type information in your system. However, if methods can now have 'var' return types, then your edits can have widespread impact and such optimizations would have to be removed.
>
> I think many people on the LDM are very interesting/open to the idea of being able to use 'var' judiciously in more places to clean the langauge up and avoid all the type-redundancy you can see. However, investigations into this have shown it is non-trivial and comes with very significant and serious engineering concerns.

None of your arguments address the primary concern that this is simply a *very costly feature*, and that we could deliver a lot more value by working on less costly features with similar end user benefits.

**Comment by Makeloft** [comment]
@CyrusNajmabadi 
In this case want to say only:
Please, don't add into C# "target typed new" because I don't want suffer from inconsistent much more.
It is true. :)

**Comment by yaakov-h** [comment]
What if we improve the IDE's code-fixes so that it can lift locals to instance variables and automatically replace `var` with the inferred type?

**Comment by Makeloft** [comment]
@yaakov-h 
Sounds better, but maybe the best solution is left side yet.
Because there are possible another types of inconsistency.

Compare
```
private IModel _model = new Model();
private IModel _model = GetModel();

private var _model = new Model();
private var _model = GetModel();

static IModel GetModel() => new AnotherModel(); // logic may be more complex

void Method()
{
    IModel _model = new Model();
    IModel _model = GetModel();

    var _model = new Model();
    var _model = GetModel();
    ...
}
```

```
private IModel _model = new Model();
private IModel _model = GetModel();

private Model _model = new(); /* some inconsistency if we want to use GetModel() method instead */

IModel GetModel() => new SecondModel();

void Method()
{
    IModel _model = new Model();
    IModel _model = GetModel();

    var _model = new Model();
    var _model = GetModel();
    ...
}
```

**Comment by MattDillard** [comment]
We're stuck in a loop!

1. I have a problem. I want solution A.
2. I sympathize, but solution A is terribly expensive. Solution B is simpler, and could come quickly.
3. I don't like solution B, because [cut & paste doesn't work | it's inconsistent]. Let's go with solution A.
4. Go to step 2.

@Makeloft Personally, I agree that using `var` for fields would look much nicer (although I'd object to using `var` for properties & method return values). I have to believe the language design team when they say that it would be a very expensive solution, though.

Given that, I'd much rather have a reasonable-but-not-ideal solution to the problem than sacrifice the solution on the altar of consistency.

**Comment by Joe4evr** [comment]
The language team has made it clear that they don't care about it being inconsistent.

**Comment by Makeloft** [comment]
@Joe4evr
Sorry, but it is clear in theory, but on practice Cut&Paste inconsistent is very annoying on workflow...

**Comment by CyrusNajmabadi** [comment]
> Just don't use both solutions! 

I think there's a lot of value in target-typed-new.  I don't see why we wouldn't want to provide it.

**Comment by CyrusNajmabadi** [comment]
> The language team has made it clear that they don't care about it being inconsistent.

:-P

As with most things, the LDM is *pragmatic*.  Consistency is something considered.  But it's not something that just definitively wins out over everything.

**Comment by CyrusNajmabadi** [comment]
> Because first is so much cost but second is so much inconsistent.

I don't see how the second is inconsistent.  It's just a new way to write things based on a contextual type.  i.e. the same way that we allow tuples to not need names provided if we can determine them contextually:


```c#
    (int x, int y) Foo()
    {
        return (0, 0);
    }
```

There was no need to name those tuple fields in ```(0, 0)``` because they can be inferred from context.  Same with "default".  You can write "default" in many contexts and the compiler can infer that it should be default(SomeType) based on that context.

THe compiler is already using contextual typing today for several features.  The idea is to extend it to constructors.  If the compiler already has a known class as the contextual type for some expression, then allow eliding the name of that class when instantiating it, just like you can elide the name of it when using 'default'.

**Comment by Makeloft** [comment]
@CyrusNajmabadi 
Sorry that you decide inside team but don't listen developer who use your tools very often. Mostly I like it but some things are annoying for long time. Now you want to do one of this things more annoying.

**Comment by CyrusNajmabadi** [comment]
> @Makeloft Personally, I agree that using var for fields would look much nicer 

Agreed.  IMO we're at an impasse.  I've explained that there are downsides and costs that would cause this feature to be deprioritized against other ones.  I'm not sure what additional information can be provided at this point.

--

@Makeloft if you are really passionate about this feature, then i would totally recommend you going and learning about the compiler/IDE innards in order to create a prototype of it yourself.  It would be valuable to understand the compiler's different phases and how it builds a semantic understanding of the declarations of your code in a highly concurrent manner.

The main difficulty here is that method-body-analysis could always depend on that ability to understand declarations without recursion.  Now, you'll need to design things such that method body analysis can then try to understand a declaration which itself needs method body analysis.  

Good luck! :)

**Comment by CyrusNajmabadi** [comment]
> Sorry that you decide inside team but don't listen developer who use your tools very often. 

I have listened.  I've even *explicitly*:

> I think many people on the LDM are very interesting/open to the idea of being able to use 'var' judiciously in more places to clean the langauge up and avoid all the type-redundancy you can see. However, investigations into this have shown it is non-trivial and comes with very significant and serious engineering concerns.

To repeat: the team is totally ok with the idea of using "var" for things like fields.  The problem is not one of desire.  The problem is one of *cost*.  And there's no magical wand i can wave to make those costs cheap :-/

--

> Now you want to do one of this things more annoying.

I honestly don't see how adding target-type-new makes things more annoying.  It's just an option to make it possible for you to elide a type in contexts where the type is known.  Just as we made that possible with ```default(SomeType) -> default``` in c# 7.  I don't think it was annoying to add that. It just made it so you could type less.  If you don't want to use it, you don't have to.  But if you don't like the type repetition then this can be quite nice to simplify things.

**Comment by Makeloft** [comment]
@CyrusNajmabadi 
Analyzers will suggest by default to use this new short form, and people will use it, but this small annoying procedure with 'var' at the current moment, will stay much more annoying with target-typed-new.

For me minuses from insertion of this feature greater then pluses.

**Comment by Makeloft** [comment]
And when this feature will be released we lost last hope to get in the far feature "true" type inference for any members, not only for fields.

**Comment by CyrusNajmabadi** [comment]
> And when this feature will be released we lost last hope to get in the far feature "true" type inference for any members, not only for fields.

FWIW, this isn't true.  The LDM would still be open to 'var' if and when it makes sense from an engineering/timeline perspective.

**Comment by Makeloft** [comment]
@CyrusNajmabadi 
As I understand you use this repo [ https://github.com/dotnet/roslyn ] for compiler , is it right?
How much cost acceptable for you for full type inference feature [may write to mail: makeman@tut.by]?

**Comment by CyrusNajmabadi** [comment]
> As I understand you use this repo [ https://github.com/dotnet/roslyn ] for compiler , is it right?

Yup!

> How much cost acceptable for you for full type inference feature 

Not really certain.  One of my ways of bucketing is to relate to tshirt sizes.  i.e. small, medium, or large (and sometimes xx-large :D).  If this were small i know i would take it.  Medium... maybe.  Large or xx-large, no.

**Comment by Makeloft** [comment]
Close to save your time, guys. Thank for participate.
